### PR TITLE
Add image pruning

### PR DESCRIPTION
**Pull request by ncdc**

### CODE REVIEW

File: pkg/image/prune/controller.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	c.addPodsToGraph(g, pods)
+	c.addReplicationControllersToGraph(g, rcs)
+	c.prune(g)
+
+	return nil
+}
+
+const minimumImageAgeInMinutesToPrune = 60
+
+func (c *PruningController) addImagesToGraph(g graph.Graph, images *imageapi.ImageList) {
```
Comment (smarterclayton) [inline pkg/image/prune/controller.go]: style note, this really doesn't have anything to do with pruning controller.  Just make it a method addImagesToGraph

Comment (ncdc) [inline pkg/image/prune/controller.go]: I considered that, but I need c.registryURL. Do you want me to pass that in?

File: pkg/image/prune/controller.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	allNodes := g.NodeList()
+	for i := range allNodes {
+		node := allNodes[i]
+		imageNode, ok := node.(*graph.ImageNode)
+		if !ok {
+			continue
+		}
+
+		onlyWeakReferences := true
+		imageStreamPredecessors := []*imageapi.ImageStream{}
```
Comment (smarterclayton) [inline pkg/image/prune/controller.go]: Why are you tracking this?

File: pkg/image/prune/controller.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+func (c *PruningController) addReplicationControllersToGraph(g graph.Graph, rcs *kapi.ReplicationControllerList) {
+	for i := range rcs.Items {
+		rc := &rcs.Items[i]
+		rcNode := graph.ReplicationController(g, rc)
+		c.addPodSpecToGraph(g, &rc.Spec.Template.Spec, rcNode)
+	}
+}
+
+func (c *PruningController) prune(g graph.Graph) error {
```
Comment (smarterclayton) [inline pkg/image/prune/controller.go]: This is just a method that takes a graph and an Image client.

File: pkg/image/prune/controller.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+				}
+			}
+		}
+	}
+}
+
+func (c *PruningController) addPodsToGraph(g graph.Graph, pods *kapi.PodList) {
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+		if pod.Status.Phase != kapi.PodRunning && pod.Status.Phase != kapi.PodPending {
```
Comment (smarterclayton) [inline pkg/image/prune/controller.go]: You may want to apply a time factor here.

File: pkg/image/prune/controller.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/image/prune/controller.go]: Create 

type PruneImagesAlgorithm struct {
  // things needed for prune
}

func (PruneImagesAlgorithm) Run() {}

I'd like the traversal code be separate from the graph setup code separate from the actual "things to delete".  For instance, creating the image graph is one step NewPruneImagesAlgorithm(), you call Run(fn func(Image)) or something and return a count of the number of images affected.

----- Original Message -----

File: pkg/image/prune/controller.go (hunk 2 above)
Comment (ncdc) [inline pkg/image/prune/controller.go]: I have more locally - it's to update stream.status.tags to remove images deleted by the pruning.

Comment (smarterclayton) [inline pkg/image/prune/controller.go]: Hrm - would prefer that you find a way to keep the coupling low between various parts here.  Doing a second graph traversal isn't that bad.

----- Original Message -----

File: pkg/image/prune/controller.go (hunk 1 above)
Comment (ncdc) [inline pkg/image/prune/controller.go]: so would you do this?

```go
type ImagePruner struct {
  g graph.Graph
  images client.ImageInterface
  // ...
}
```

And what would the func that takes an Image that you pass to Run do?

Comment (smarterclayton) [inline pkg/image/prune/controller.go]: ----- Original Message -----
> > +		return fmt.Errorf("error listing all replication controllers: %v", err)
> > +	}
> > +
> > +	g := graph.New()
> > +	c.addImagesToGraph(g, images)
> > +	c.addImageStreamsToGraph(g, streams)
> > +	c.addPodsToGraph(g, pods)
> > +	c.addReplicationControllersToGraph(g, rcs)
> > +	c.prune(g)
> > +
> > +	return nil
> > +}
> > +
> > +const minimumImageAgeInMinutesToPrune = 60
> > +
> > +func (c *PruningController) addImagesToGraph(g graph.Graph, images
> > *imageapi.ImageList) {
> 
> so would you do this?
> 
> ```go
> type ImagePruner struct {
>   g graph.Graph
>   images client.ImageInterface
>   // ...
> }
> ```
> 
> And what would the func that takes an Image that you pass to Run do?
> 

I don't think the Pruner needs to know about client.ImageInterface - the Function you pass to run would do.  That way you can dry run without having to change the algorithm.

So a set of stages:

1. Need to load and fetch objects and add them to a graph
2. Identify any images that aren't referenced by the graph
3. Delete those images

But you could replace #3 with print the images out.  Or replace #3 with a function that deletes the images AND removes them from an IR.  Or removes them from the IR, and then deletes them later.

It's important to keep that last step separate from the first two.

Comment (ncdc) [inline pkg/image/prune/controller.go]: Something along these lines?

```go
type ImagePruner interface {
  ListPrunableImages() []*imageapi.Image
}

type imagePruner struct {
  g graph.Graph
}

func NewImagePruner(images client.ImageInterface, streams ...) ImagePruner {
  g := graph.New()
  addImagesToGraph(g, images)
  // ...
  return &imagePruner{g:g}
}
```

Comment (ncdc) [inline pkg/image/prune/controller.go]: Or should NewImagePruner take the actual lists of images, streams, pods, rcs that have been retrieved elsewhere?

File: pkg/image/prune/orphanedimagefinder.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		}
+	}
+}
+
+func addPodsToGraph(g graph.Graph, pods *kapi.PodList, registryURL string) {
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+
+		age := util.Now().Sub(pod.CreationTimestamp.Time)
+		if age.Minutes() < minimumAgeInMinutesToPrune {
```
Comment (smarterclayton) [inline pkg/image/prune/orphanedimagefinder.go]: This is something that should be on a pruneAlgorithm object (the value) - if you do that, you can move registryURL there

File: pkg/image/prune/orphanedimagefinder.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	}
+}
+
+func addPodSpecToGraph(g graph.Graph, spec *kapi.PodSpec, predecessor gonum.Node, registryURL string) {
+	for j := range spec.Containers {
+		container := spec.Containers[j]
+		if !strings.HasPrefix(container.Image, registryURL) {
+			continue
+		}
+		parts := strings.Split(container.Image, "@")
```
Comment (smarterclayton) [inline pkg/image/prune/orphanedimagefinder.go]: DockerImageReference

File: pkg/image/prune/orphanedimagefinder.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+func addReplicationControllersToGraph(g graph.Graph, rcs *kapi.ReplicationControllerList, registryURL string) {
+	for i := range rcs.Items {
+		rc := &rcs.Items[i]
+		rcNode := graph.ReplicationController(g, rc)
+		addPodSpecToGraph(g, &rc.Spec.Template.Spec, rcNode, registryURL)
+	}
+}
+
+func (f *orphanedImageFinder) Find() []*imageapi.Image {
```
Comment (smarterclayton) [inline pkg/image/prune/orphanedimagefinder.go]: I think I prefer the iterator style approach which removes the need to have a list of all images to prune - but you could expose both (where Find() invokes Run(fn func()))

Comment (ncdc) [inline pkg/image/prune/orphanedimagefinder.go]: I was thinking of using Find to return a list of images that would be pruned to assist with the dry run. I guess we could go back to just Run(fn func(...)) and swap out fn to a no-op impl for dry run that would print out the images. Is that what you were thinking?

Comment (smarterclayton) [inline pkg/image/prune/orphanedimagefinder.go]: Yeah

----- Original Message -----

Comment (ncdc) [inline pkg/image/prune/orphanedimagefinder.go]: Cool, on it!

File: pkg/image/prune/controller.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/image/prune/controller.go]: Yeah, that's probably the easiest way.  Yo could configure your own graph or call the NewMethod.

----- Original Message -----

File: pkg/image/prune/orphanedimagefinder.go (hunk 7 above)
Comment (ncdc) [inline pkg/image/prune/orphanedimagefinder.go]: How about updating the image streams to remove pruned images from status.tags... should Run() return the subgraph of pruned images, and another function can be used to update the image streams? Or should Run() handle that too?

Comment (smarterclayton) [inline pkg/image/prune/orphanedimagefinder.go]: Run should be passing enough info to the function that you can do what you need - i.e., maybe you get passed the Node instead of the image and you can look the image up in the graph to find its incoming.  Or you can just pass two arguments to the func - the image, and the image streams that reference it.

----- Original Message -----

Comment (ncdc) [inline pkg/image/prune/orphanedimagefinder.go]: Ok, cool, I can pass the image and the streams.

**Comment by bparees** [comment]
Can you summarize somewhere the overall pruning strategy?  what gets pruned, what is sufficient to prevent something from being pruned, etc?

It looks like any image that's the current image for a tag won't be pruned, but it's less clear how running pods and RCs are intended to factor in (looks like they keep images around).  What about a BuildConfig that references a particular ImageStreamImage?  Should also note that pruning old images could make it impossible to rerun a previous build since the exact image it used won't be available anymore (not suggesting we shouldn't prune those, but it's a ramification that should be doc'd)

**Comment by ncdc** [comment]
The description is here: https://trello.com/c/dWeQNKRL

BuildConfigs are not taken into account at the moment. Should they be?

**Comment by bparees** [comment]
seems like they should be.  Again, ImageStream and ImageStreamTag references from BuildConfigs are presumably already covered by the "image that represents the the current state of a tag in a stream" rule, but at a minimum an ImageStreamImage reference from a BuildConfig should prevent pruning.

What we want to do about Builds that consumed a particular Image is less clear to me, but it seems like at least an option to not prune those images would be a good idea.  or better, not pruning them by default.

**Comment by ncdc** [comment]
Let's discuss build configs soon. 

If we have a plan for pruning builds, I'm fine including them in the criteria for image pruning.

File: pkg/image/prune/imagepruner.go [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+		dc := &dcs.Items[i]
+		dcNode := graph.DeploymentConfig(g, dc)
+		addPodSpecToGraph(g, &dc.Template.ControllerTemplate.Template.Spec, dcNode, algorithm)
+	}
+}
+
+// addBuildConfigsToGraph adds build configs to the graph.
+//
+// Edges are added to the graph from each build config to the image specified by its strategy.from.
+func addBuildConfigsToGraph(g graph.Graph, bcs *buildapi.BuildConfigList, algorithm pruneAlgorithm) {
```
Comment (bparees) [inline pkg/image/prune/imagepruner.go]: addBuild* methods look right to me.

File: pkg/image/prune/imagepruner.go [hunk 9]
```diff
@@ -0,0 +1,744 @@
...
+type imagePruner struct {
+	g         graph.Graph
+	algorithm pruneAlgorithm
+}
+
+var _ ImagePruner = &imagePruner{}
+
+/*
+NewImagePruner creates a new ImagePruner.
+
```
Comment (pmorie) [inline pkg/image/prune/imagepruner.go]: This is the right place to have a high-level comment describing the lack of a delete API in the registry and explaining why this is factored like it is.

File: Godeps/_workspace/src/github.com/docker/distribution/registry/storage/layerstore.go [hunk 10]
```diff
@@ -52,6 +52,18 @@ func (ls *layerStore) Fetch(dgst digest.Digest) (distribution.Layer, error) {
...
 }
 
+func (ls *layerStore) Delete(dgst digest.Digest) error {
+	lp, err := ls.linkPath(dgst)
+	if err != nil {
+		return err
+	}
+
+	// strip off /link at the end
+	lp = lp[:len(lp)-5]
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/docker/distribution/registry/storage/layerstore.go]: Do you not want to check the suffix before doing this?  I recently got in trouble making an assumption and doing the same thing.

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 11]
```diff
@@ -18,24 +19,167 @@ import (
...
+
+	return router
+}
+
+// DeleteLayersRequest is a mapping from layers to the image repositories that
+// reference them. Below is a sample request:
+//
+// {
+//   "layer1": ["repo1", "repo2"],
+// 	 "layer2": ["repo1", "repo3"],
```
Comment (pmorie) [inline pkg/cmd/dockerregistry/dockerregistry.go]: Is the spacing in this comment what you want?

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 12]
```diff
@@ -18,24 +19,167 @@ import (
...
+// 	 ...
+// }
+type DeleteLayersRequest map[string][]string
```
Comment (pmorie) [inline pkg/cmd/dockerregistry/dockerregistry.go]: It would be good to describe what the keys and values in this map are.

File: pkg/cmd/dockerregistry/dockerregistry.go [hunk 13]
```diff
@@ -18,24 +19,167 @@ import (
...
+}
+
+// AddStream adds an image stream reference to the layer.
+func (r DeleteLayersRequest) AddStream(layer, stream string) {
+	r[layer] = append(r[layer], stream)
+}
+
+type DeleteLayersResponse struct {
+	Result string
+	Errors map[string][]string
```
Comment (pmorie) [inline pkg/cmd/dockerregistry/dockerregistry.go]: Describe format of error

File: Godeps/_workspace/src/github.com/docker/distribution/registry/storage/layerstore.go (hunk 10 above)
Comment (ncdc) [inline Godeps/_workspace/src/github.com/docker/distribution/registry/storage/layerstore.go]: Probably safer. `linkPath` currently has `/link` at the end, but i can guard against it not being there.

File: pkg/cmd/dockerregistry/dockerregistry.go (hunk 12 above)
Comment (ncdc) [inline pkg/cmd/dockerregistry/dockerregistry.go]: You mean the comment I have above? :smile:

File: pkg/api/graph/graph.go [hunk 14]
```diff
@@ -9,7 +9,7 @@ import (
 )
 
 type Node struct {
-	concrete.Node
+	graph.Node
```
Comment (smarterclayton) [inline pkg/api/graph/graph.go]: Why this change?

File: pkg/image/prune/imagepruner.go [hunk 15]
```diff
@@ -0,0 +1,736 @@
...
+//
+// The body of the request is JSON, and it is a map[string][]string, with each
+// key being a layer, and each value being a list of Docker image repository
+// names referenced by the layer.
+func DeletingLayerPruneFunc(registryClient *http.Client) LayerPruneFunc {
+	return func(registryURL, repoName, layer string) error {
+		glog.V(4).Infof("Pruning registry %q, repo %q, layer %q", registryURL, repoName, layer)
+
+		//TODO https
+		req, err := http.NewRequest("DELETE", fmt.Sprintf("http://%s/admin/%s/layers/%s", registryURL, repoName, layer), nil)
```
Comment (smarterclayton) [inline pkg/image/prune/imagepruner.go]: Why not make the protocol part of something higher level (abstract into a method of a class you invoke)?

**Comment by smarterclayton** [comment]
Not a lot of comments

File: pkg/api/graph/graph.go (hunk 14 above)
Comment (ncdc) [inline pkg/api/graph/graph.go]: I thought it was needed for something I was working on a while ago, but it's not - I can undo it.

File: pkg/image/prune/imagepruner.go (hunk 15 above)
Comment (ncdc) [inline pkg/image/prune/imagepruner.go]: I assume we'll need to try both https and http, so this was really just a reminder to get that in there.

**Comment by ncdc** [comment]
@smarterclayton did you look at my changes to Godeps?

**Comment by smarterclayton** [comment]
Yes, nothing particularly jumped out as bad.

**Comment by ncdc** [comment]
Cool, thanks. I'm working on a bit more cleanup (to hopefully get rid of the retries when updating the image streams).

**Comment by ncdc** [comment]
[test]

**Comment by ncdc** [comment]
@smarterclayton ok to merge so QE can start testing? or perhaps a fork ami?

**Comment by derekwaynecarr** [comment]
This was lgtm as well.  Forgot to comment as such.

**Comment by smarterclayton** [comment]
Go ahead and merge, there really no negative impact.

**Comment by ncdc** [comment]
Ok, hopefully my changes to the registry to support pruning don't have any adverse effects. I've been testing locally and haven't had any issues pushing images.

[merge]
